Computed map properties

nattridgell opened the pull request:
- adds a list of properties that take length values
  to ComputedStylePropertyMap
  - adds a check for keywords
    NOTE: this may have to change / move around
  - adds some basic tests for length and keyword values
    for the map
    NOTE: These tests are currently being skipped as
     parsing is unimplemented.
  - adds `auto` keyword to KeywordValue
  - alter KeywordValue related tests as appropriate

Commit:
Add keywords and properties to computed map

 - adds a list of properties that take length values
   to ComputedStylePropertyMap
 - adds a check for keywords
   NOTE: this may have to change / move around
 - adds some basic tests for length and keyword values
   for the map
   NOTE: These tests are currently being skipped as
         parsing is unimplemented.

Commit:
Adds `auto` to KeywordValue

 - adds auto to the possible KeywordValues
 - adjusts tests in ComputedStylePropertyMap for KeywordValues

nattridgell commented:
Now that we are temporarily returning to the div-soup approach for keywords, I have re-opened this PR, with the small change/comment addressed from the previous code review.

wilddamon commented on src/keyword-value.js:
```diff
@@ -22,7 +22,7 @@
   }
 
-  KeywordValue.StyleValueKeyword = ['initial', 'inherit', 'revert', 'unset'];
+  KeywordValue.StyleValueKeyword = ['auto', 'initial', 'inherit', 'revert', 'unset'];
```
Perhaps this should be GlobalKeyword, and other things that refer to it for parsing/validation purposes should also refer to "isGlobalKeyword" and such. I think the keywords for particular types should go in the classes used for those purposes. It's probably still worth discussion of what the structure should look like for something like "speech-rate" which takes 

<number> | x-slow | slow | medium | fast | x-fast | faster | slower | inherit

since it doesn't look like there should be a SpeechRateValue class...

nattridgell replied:
Cool beans :-)  I'll re-mock it up :D

rjwright replied:
> It's probably still worth discussion of what the structure should look like for something like "speech-rate" which takes
> 
> | x-slow | slow | medium | fast | x-fast | faster | slower | inherit
> 
> since it doesn't look like there should be a SpeechRateValue class...

I think having a SpeechRateValue class just for the purpose of basing a  SpeechRateKeywordValue class would be OK. But there are many things you could do; we should discuss it.

rjwright replied:
Wait, speech-rate can also take a number. That changes things.

nattridgell replied:
As we are going for the keyword soup approach, I don't think "isGlobalKeyword" is as appropriate as before, nor do I think that the following comment is still applicable.  I stand to be corrected though.
"I think the keywords for particular types should go in the classes used for those purposes."

So, I'll just let this be for now, unless otherwise advised.

rjwright commented on src/computed-style-property-map.js:
```diff
@@ -78,15 +103,28 @@
 
     switch (property) {
```
This switch is getting large. Do we need it? ComputedStylePropertyMap.get throws for unsupported properties. Can we just do

```
var value = this.get(property);
if (Array.isArray(value)) {
    // TODO: Handle shorthands
    throw new TypeError('Not implemented yet');
} else {
    return value ? [value] : [];
}
```

here?

nattridgell replied:
Sounds like a good idea for now, might need to restructure later on.  It is getting quite unwieldy, I agree.

Commit:
Update ComputedStylePropertyMap.getAll

Change from having two independent switch statements
in get / getAll, to just using the get method inside getAll.

nattridgell commented on src/computed-style-property-map.js:
```diff
@@ -56,6 +56,31 @@
+      case 'word-spacing':
+        throw new TypeError('Not implemented yet');
+
+      // These properties take a number or SimpleLength
```
This comment should be removed.

nattridgell commented:
Irrelevant with code restructure of computed style property map.
